src/Monotonic.cpp

```diff
@@ -233,71 +362,58 @@ class MonotonicVisitor : public IRVisitor {
+        ConstantInterval ra = result;
         op->b.accept(this);
-        Monotonic rb = result;
+        ConstantInterval rb = result;
         result = unify(ra, rb);
     }
 
     void visit(const Or *op) override {
         op->a.accept(this);
-        Monotonic ra = result;
+        ConstantInterval ra = result;
         op->b.accept(this);
-        Monotonic rb = result;
+        ConstantInterval rb = result;
         result = unify(ra, rb);
     }
 
     void visit(const Not *op) override {
         op->a.accept(this);
-        result = flip(result);
+        result = negate(result);
     }
 
     void visit(const Select *op) override {
-        op->condition.accept(this);
-        Monotonic rcond = result;
-
-        op->true_value.accept(this);
-        Monotonic ra = result;
-        op->false_value.accept(this);
-        Monotonic rb = result;
-        Monotonic unified = unify(ra, rb);
-
-        if (rcond == Monotonic::Constant) {
-            result = unified;
-            return;
-        }
-
-        bool true_value_ge_false_value = can_prove(op->true_value >= op->false_value);
-        bool true_value_le_false_value = can_prove(op->true_value <= op->false_value);
-
-        bool switches_from_true_to_false = rcond == Monotonic::Decreasing;
-        bool switches_from_false_to_true = rcond == Monotonic::Increasing;
-
-        if (true_value_ge_false_value &&
-            true_value_le_false_value) {
-            // The true value equals the false value.
-            result = ra;
-        } else if ((unified == Monotonic::Increasing || unified == Monotonic::Constant) &&
-                   ((switches_from_false_to_true && true_value_ge_false_value) ||
-                    (switches_from_true_to_false && true_value_le_false_value))) {
-            // Both paths increase, and the condition makes it switch
-            // from the lesser path to the greater path.
-            result = Monotonic::Increasing;
-        } else if ((unified == Monotonic::Decreasing || unified == Monotonic::Constant) &&
-                   ((switches_from_false_to_true && true_value_le_false_value) ||
-                    (switches_from_true_to_false && true_value_ge_false_value))) {
-            // Both paths decrease, and the condition makes it switch
-            // from the greater path to the lesser path.
-            result = Monotonic::Decreasing;
+        // The result is the unified bounds, added to the "bump" that happens when switching from true to false.
+        if (op->type.is_scalar()) {
+            op->condition.accept(this);
+            ConstantInterval rcond = result;
+
+            op->true_value.accept(this);
+            ConstantInterval ra = result;
+            op->false_value.accept(this);
+            ConstantInterval rb = result;
+            ConstantInterval unified = unify(ra, rb);
+
+            // TODO: How to handle unsigned values?
+            Expr delta = simplify(op->true_value - op->false_value);
```

Comment:
Calling simplify in this visitor is probably a bad idea. Can it be avoided?

Reply:
I think it might be necessary. FWIW the old logic used 2 can_prove calls on similar expressions, so this seems like an improvement...

Reply:
Fair enough, I guess you need it to fold to a constant.